src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp

```diff
@@ -24,23 +24,516 @@
+  G1PreBarrierStubC2* const stub = new (Compile::current()->comp_arena()) G1PreBarrierStubC2(node);
+  if (!Compile::current()->output()->in_scratch_emit_size()) {
+    barrier_set_state()->stubs()->append(stub);
+  }
+  return stub;
+}
+
+void G1PreBarrierStubC2::initialize_registers(Register obj, Register pre_val, Register thread, Register tmp1, Register tmp2) {
+  _obj = obj;
+  _pre_val = pre_val;
+  _thread = thread;
+  _tmp1 = tmp1;
+  _tmp2 = tmp2;
+}
+
+Register G1PreBarrierStubC2::obj() const {
+  return _obj;
+}
+
+Register G1PreBarrierStubC2::pre_val() const {
+  return _pre_val;
+}
+
+Register G1PreBarrierStubC2::thread() const {
+  return _thread;
+}
+
+Register G1PreBarrierStubC2::tmp1() const {
+  return _tmp1;
+}
+
+Register G1PreBarrierStubC2::tmp2() const {
+  return _tmp2;
+}
+
+void G1PreBarrierStubC2::emit_code(MacroAssembler& masm) {
+  G1BarrierSetAssembler* bs = static_cast<G1BarrierSetAssembler*>(BarrierSet::barrier_set()->barrier_set_assembler());
+  bs->generate_c2_pre_barrier_stub(&masm, this);
+}
+
+G1PostBarrierStubC2::G1PostBarrierStubC2(const MachNode* node) : G1BarrierStubC2(node) {}
+
+bool G1PostBarrierStubC2::needs_barrier(const MachNode* node) {
+  return (node->barrier_data() & G1C2BarrierPost) != 0;
+}
+
+G1PostBarrierStubC2* G1PostBarrierStubC2::create(const MachNode* node) {
+  G1PostBarrierStubC2* const stub = new (Compile::current()->comp_arena()) G1PostBarrierStubC2(node);
+  if (!Compile::current()->output()->in_scratch_emit_size()) {
+    barrier_set_state()->stubs()->append(stub);
+  }
+  return stub;
+}
+
+void G1PostBarrierStubC2::initialize_registers(Register thread, Register tmp1, Register tmp2) {
+  _thread = thread;
+  _tmp1 = tmp1;
+  _tmp2 = tmp2;
+}
+
+Register G1PostBarrierStubC2::thread() const {
+  return _thread;
+}
+
+Register G1PostBarrierStubC2::tmp1() const {
+  return _tmp1;
+}
+
+Register G1PostBarrierStubC2::tmp2() const {
+  return _tmp2;
+}
+
+void G1PostBarrierStubC2::emit_code(MacroAssembler& masm) {
+  G1BarrierSetAssembler* bs = static_cast<G1BarrierSetAssembler*>(BarrierSet::barrier_set()->barrier_set_assembler());
+  bs->generate_c2_post_barrier_stub(&masm, this);
+}
+
+void* G1BarrierSetC2::create_barrier_state(Arena* comp_arena) const {
+  return new (comp_arena) G1BarrierSetC2State(comp_arena);
+}
+
+int G1BarrierSetC2::get_store_barrier(C2Access& access, C2AccessValue& val) const {
+  int barriers = G1C2BarrierPre | G1C2BarrierPost;
+  if (!access.is_parse_access()) {
+    // Only support for eliding barriers at parse time for now.
+    return barriers;
+  }
+  GraphKit* kit = (static_cast<C2ParseAccess&>(access)).kit();
+  Node* adr = access.addr().node();
+  uint adr_idx = kit->C->get_alias_index(access.addr().type());
+  assert(adr_idx != Compile::AliasIdxTop, "use other store_to_memory factory" );
+
+  if (g1_can_remove_pre_barrier(kit, &kit->gvn(), adr, access.type(), adr_idx)) {
+    barriers ^= G1C2BarrierPre;
```

Comment:
Is it possible to rewrite this method to remove `^=`? This method first construct both pre/post barriers and remove if needed. I wonder if the logic will be cleaner if we track if pre/post barrier can be removed using two `bool` and construct the actual barrier in the end using two `bool` -- this way, we can avoid mutating the `barriers` var.

Reply:
Thanks for the suggestion, I will try it out.

Reply:
Is [this](https://github.com/robcasloz/jdk/compare/JDK-8334060-g1-late-barrier-expansion...robcasloz:jdk:JDK-8334060-g1-late-barrier-expansion-alberts-barriers-suggestion) what you had in mind? I don't have a strong opinion, but if you think the change improves readability I am happy to merge it.

Reply:
Yes. 

Two nits: add `can_` to those two bools and unpack the final return expr, sth like:

```
int barriers = 0;

if (!can_remove_pre...) {
  barriers |= pre;
}
if (!can_remove_post...) {
  barriers |= post;
}

return barriers;
```

Reply:
Thanks, I will do some testing before merging.

Reply:
Done (commit d722d4c7c1534794aaa38d54ecf7a4c12b158e84).